MG-2041 - Add user's guide blogpost (#195)
absmach/magistrala-docs: docs/blogs/user-guide.md

3. **Enhanced Security**: With fine-grained access control, Magistrala ensures that only authorized entities can access sensitive data and system functionalities.

4. **Insightful Monitoring**: Extensive logging and instrumentation support allows for detailed system analysis, facilitating efficient troubleshooting and performance optimization.

5. **Efficient Deployment**: Leveraging container-based deployment with Docker, Magistrala enables quick and consistent setup across various environments.

6. **Robust Data Handling**: The platform's efficient data ingestion, processing, and storage capabilities support real-time analytics and informed decision-making.

7. **Scalability and Reliability**: Designed to be scalable and distributed, Magistrala maintains high availability even as demands increase.

8. **Multi-Tenancy Support**: Multiple users can operate independently within the same platform, enhancing flexibility and resource utilization.

9. **Customizable and Extensible**: Magistrala's architecture allows for easy customization and extension to meet specific business needs.

Certainly! Here's a more professional phrasing:

> **Please note:** Magistrala offers a variety of services, each detailed comprehensively in the [architecture documentation](../architecture.md).

Reviewer: Since this will be published to medium use absolute links https://docs.magistrala.abstractmachines.fr/architecture/

Magistrala's Command Line Interface (CLI) provides a powerful way to interact with the platform. Here are some basic operations with detailed explanations:

- Create a user:

```bash
magistrala-cli users create John johndoe@example.com 12345678
```

   This command creates a new user named John with the email `johndoe@example.com` and password `12345678`.

- Generate an access token:

```bash
magistrala-cli users token johndoe@example.com 12345678
```

   This generates an access token for the user, which is required for authentication in subsequent operations.

   The access token can be set to an environmental variable to allow for easy use:

```bash
ACCESSTOKEN=eyJhbGciOiJIUzUxMiIsInR5cCI6IkpXVCJ9.eyJleHAiOjE3MTgxOTI2MTIsImlhdCI6MTcxODE4OTAxMiwiaXNzIjoibWFnaXN0cmFsYS5hdXRoIiwidHlwZSI6MCwidXNlciI6ImMxYjQ3OWQwLWEzNDYtNGZiYy1hN2FkLWVkNGFiNjcxYTIyZCJ9._Qlri9FN_E2vJLZIVQzyeeg9I_ggVRi_CKS52xlUD8YEkSKWkTlmr02WlKdhvR-aZzaDyudICLtCPtadMCGCzg
```

- Create a domain (workspace):

```bash
magistrala-cli domain create workspace alias $ACCESSTOKEN
```

   This creates a new domain similar to a "workspace" with an alias. The $ACCESSTOKEN is the token generated in the previous step.

> 

Reviewer: Add step to login to domain

- Create a thing (device):

```bash
magistrala-cli things create '{"name":"Sensor1", "metadata":{"units":"cm"}}' $ACCESSTOKEN
```

   This creates a new thing named "Sensor1" with metadata specifying its units as centimeters.

- Create a channel:

```bash
magistrala-cli channels create '{"name":"Distance"}' $ACCESSTOKEN
```

   This creates a new channel named "Distance" which can be used for communication.

- Connect a thing to a channel:

```bash
magistrala-cli things connect $THINGID $CHANNELID $ACCESSTOKEN 
```

   This connects the previously created thing to the channel. $THINGID and $CHANNELID are environment variables used to store the Thing and Channel IDs, similar to how the access token is stored. These IDs are generated when the Thing and Channel are created.

- Send a message:

```bash

> magistrala-cli messages send $CHANNELID '[{"bn":"Dev1","n":"temp","v":20}, {"n":"hum","v":40}]' <thing_secret>

Reviewer: Add instruction on how to get thing secret

```

   This sends a message to the specified channel. The message contains temperature and humidity readings. Replace <thing_secret> with the secret of the thing.

- Read messages:

```bash
magistrala-cli messages read $CHANNELID $ACCESSTOKEN
```

   This reads messages from the specified channel.

You can also update a Thing's tags, secret, and metadata. Here's an example of how to do that:

```bash
magistrala-cli things update secret $THINGID $THINGSECRET $ACCESSTOKEN
```

Within your domain, you have the ability to invite or assign other users as members, guests, contributors, editors, or administrators. This allows you to grant varying levels of permissions to users over different entities. To assign a user to a domain, use the following command:

```bash
magistrala-cli domains assign users member '[<user_id>]' $DOMAINID $ACCESSTOKEN
```

**Note:** Assigning a user adds them automatically to the domain. However, if invited, the user will need to accept the invitation.

There are many commands available through the CLI. For more information, run:

> ```bash

Reviewer: This should start before using `magistrala-cli`

```

To login into your workspace to generate the token with the domain ID attached:

```bash
magistrala-cli users token johndoe@example.com 12345678 $DOMAINID
```

**Note:** The access token will need to set to the environmental variable created earlier. While logged into your domain there are muliple things that can be done. Let go through some of them.

- Create a thing (device):

```bash
magistrala-cli things create '{"name":"Sensor1", "metadata":{"units":"cm"}}' $ACCESSTOKEN
```

   This creates a new thing named "Sensor1" with metadata specifying its units as centimeters.

- Create a channel:

```bash
magistrala-cli channels create '{"name":"Distance"}' $ACCESSTOKEN
```

   This creates a new channel named "Distance" which can be used for communication.

- Connect a thing to a channel:

```bash
magistrala-cli things connect $THINGID $CHANNELID $ACCESSTOKEN 
```

   This connects the previously created thing to the channel. $THINGID and $CHANNELID are environment variables used to store the Thing and Channel IDs, similar to how the access token is stored. These IDs are generated when the Thing and Channel are created.

- Send a message:

```bash

> magistrala-cli messages send $CHANNELID '[{"bn":"Dev1","n":"temp","v":20}, {"n":"hum","v":40}]' <thing_secret>

Reviewer: From the naming of the thing and channel, it is assumed it will be sending distance in cm correct this.

Something like

```json
{
  "name": "Pressure Sensor Tank A",
  "metadata": {
    "type": "sensor",
    "location": "Tank A",
    "manufacturer": "Acme Corp",
    "model": "PS-2000",
    "installation_date": "2023-05-12",
    "maintenance_due": "2024-05-12",
    "communication_protocol": "MQTT",
    "battery_level": 90
  },
  "tags": [
    "pressure",
    "tank",
    "critical"
  ]
}
```

```json
{
  "name": "Tank A Outlet Pressure",
  "metadata": {
    "unit": "bar",
    "sampling_interval": "10s",
    "data_type": "float",
    "min_value": 0,
    "max_value": 10,
    "precision": 0.01,
    "threshold_warning": 8.5,
    "threshold_critical": 9.5
  }
}
```

```json
[
  {
    "bn": "<device-serial-number>",
    "n": "Pressure_TankA_Outlet",
    "u": "bar",
    "v": 7.8
  },
  {
    "n": "BatteryLevel",
    "u": "%",
    "v": 88
  }
]
```

## Keywords

IIoT, Open Source, IoT Platform, Industrial Automation, Remote Monitoring, MQTT, CoAP, WebSocket, Device Management, Scalability, Security, Docker, Go, Multi-Tenancy, Data Processing, Real-Time Analytics, Edge Computing, Sensor Networks, Industrial IoT, Connected Devices, IoT Middleware, Protocol Bridging, Cloud Platform, IoT Security, IoT Data Management, Customer Support, Community Engagement, Open Source Community, IIoT Support, Developer Resources

[architecture]: https://docs.magistrala.abstractmachines.fr/architecture/
[authorization]: https://docs.magistrala.abstractmachines.fr/authorization/#domain-viewer-with-channel-thing
[magistrala-repo]: https://github.com/absmach/magistrala
[mosquitto-site]: https://mosquitto.org/download/
[swagger-docs]: https://docs.api.magistrala.abstractmachines.fr/
[coap-repo]: https://github.com/absmach/coap-cli/releases/tag/v0.3.3
[websocat-repo]: https://github.com/vi/websocat?tab=readme-ov-file#installation
[drasko]: https://github.com/drasko
[docs]: https://docs.magistrala.abstractmachines.fr

> [Twitter-handle]: https://x.com/absmach?t=9j4nWw8EKQJ_-IPSPbDY8g&s=09

Reviewer: ```suggestion
[Twitter-handle]: https://x.com/absmach
```

**Please note:** Magistrala offers a variety of services, each detailed comprehensively in the [architecture documentation][architecture].

## Getting Started with Magistrala

Setting up Magistrala is straightforward. With Docker (version 26.0.0 and above) and Go (version 1.21 and above) installed on your system, you can clone the official GitHub repository and have the platform up and running in no time. Here's a quick guide to get you started:

- Clone the Magistrala repository and navigate to the newly created directory:

   ```bash
   git clone https://github.com/absmach/magistrala.git
   cd magistrala
   ```

- Build and install the binaries:

   ```bash
   make && make install
   ```

**Note:** This process will compile the binaries into the `<project_root>/build` directory. If the `$GOBIN` environment variable is set, the binaries will also be copied to the `go/bin` directory.

To start Magistrala Docker services, execute the following command from the project's root directory:

   ```bash
   make run
   ```

Magistrala offers multiple interaction options, including a CLI, SDK, HTTP API, and UI, catering to different user preferences and use cases.

### Using the CLI

> Magistrala's Command Line Interface (CLI) provides a powerful way to interact with the platform. There are many commands available through the CLI. For more information, run:

Reviewer 2: ```suggestion
Magistrala's Command Line Interface (CLI) provides a powerful way to interact with the platform. There are many commands available through the CLI. For more information, refer to the [Magistrala CLI Documentation](https://docs.magistrala.abstractmachines.fr/cli/). To explore available commands and get usage details, run:
```

  -i, --insecure                 Do not check for TLS cert
  -v, --invitations-url string   Inivitations URL
  -a, --journal-url string       Journal Log URL
  -l, --limit uint               Limit query parameter (default 10)
  -m, --metadata string          Metadata query parameter
  -n, --name string              Name query parameter
  -o, --offset uint              Offset query parameter
  -r, --raw                      Enables raw output mode for easier parsing of output
  -R, --reader-url string        Reader URL
  -z, --state string             Bootstrap state query parameter
  -S, --status string            User status query parameter
  -t, --things-url string        Things service URL
  -T, --topic string             Subscription topic query parameter
  -u, --users-url string         Users service URL

Use "magistrala-cli [command] --help" for more information about a command.
```

Here are some basic operations with detailed explanations:

- Create a user:

```bash
magistrala-cli users create John johndoe@example.com 12345678
```

   This command creates a new user named John with the email `johndoe@example.com` and password `12345678`.

> - Generate an access token:

Reviewer 2: ```suggestion
- After creating a user, you need to generate an access token for them. To do so, run:
```

```bash
magistrala-cli users token johndoe@example.com 12345678
```

>    This generates an access token for the user, which is required for authentication in subsequent operations.
> 
>    The access token can be set to an environmental variable to allow for easy use:

Reviewer 2: ```suggestion
   This generates an access token and a refresh token for the created user. The access token is used for authentication in subsequent operations, while the refresh token allows you to generate a new access token when the existing access token expires. For convenience, both tokens can be set as environment variables for easy access:
```

> ```bash
> ACCESSTOKEN=eyJhbGciOiJIUzUxMiIsInR5cCI6IkpXVCJ9.eyJleHAiOjE3MTgxOTI2MTIsImlhdCI6MTcxODE4OTAxMiwiaXNzIjoibWFnaXN0cmFsYS5hdXRoIiwidHlwZSI6MCwidXNlciI6ImMxYjQ3OWQwLWEzNDYtNGZiYy1hN2FkLWVkNGFiNjcxYTIyZCJ9._Qlri9FN_E2vJLZIVQzyeeg9I_ggVRi_CKS52xlUD8YEkSKWkTlmr02WlKdhvR-aZzaDyudICLtCPtadMCGCzg

Reviewer 2: ```suggestion
```bash
ACCESSTOKEN=eyJhbGciOiJIUzUxMiIsInR5cCI6IkpXVCJ9.eyJleHAiOjE3MTgxOTI2MTIsImlhdCI6MTcxODE4OTAxMiwiaXNzIjoibWFnaXN0cmFsYS5hdXRoIiwidHlwZSI6MCwidXNlciI6ImMxYjQ3OWQwLWEzNDYtNGZiYy1hN2FkLWVkNGFiNjcxYTIyZCJ9._Qlri9FN_E2vJLZIVQzyeeg9I_ggVRi_CKS52xlUD8YEkSKWkTlmr02WlKdhvR-aZzaDyudICLtCPtadMCGCzg

REFRESHTOKEN=eyJhbGciOiJIUzUxMiIsInR5cCI6IkpXVCJ9.eyJleHAiOjE3MjU3MDM4NTUsImlhdCI6MTcyNTYxNzQ1NSwiaXNzIjoibWFnaXN0cmFsYS5hdXRoIiwidHlwZSI6MSwidXNlciI6IjM0YzNhMDhjLTBkMzEtNGU3Ny1iNjBlLTA0OTBiOTc5NzQyMSJ9.v9tFiX2NKiykg1SmmwNQrSDlcsNrYp7isF0kqTzyo8wiJuYfvO7GwCKNb1aGPx6uR6_VYGQlfN6aoxAKP_42WQ
```

```

> **Note:** The token last for 15 minutes before expiring and will need the user to get a new one.

Reviewer 2: ```suggestion
**Note:** The access token is valid for 15 minutes before it expires. After this period, the user will need to use the refresh token to obtain a new access token.
```

```

   This connects the previously created thing to the channel. $THINGID and $CHANNELID are environment variables used to store the Thing and Channel IDs, similar to how the access token is stored. These IDs are generated when the Thing and Channel are created.

- Send a message:

```bash
magistrala-cli messages send $CHANNELID '[{"bn": "DS-5000-AL1-001", "n": "Distance_AssemblyLine1", "u": "mm","v": 152.3}, { "n": "BatteryLevel", "u": "%", "v": 95}, { "n": "SignalStrength", "u": "dBm", "v": -65}]' <thing_secret>
```

   This sends a message to the specified channel. The message contains temperature and humidity readings. Replace <thing_secret> with the secret of the thing.

**Note:** The secret for the thing can be specified during its creation. If not provided, it will be automatically generated by the platform. To view the generated secret, please access the profile as follows:

```bash
magistrala-cli things get $THINGID $ACCESSTOKEN
```

Please note that appropriate access control is required to view the thing's credentials. For more information, please refer to the [documentation][Authorization].

- Read messages:

```bash

> magistrala-cli messages read $CHANNELID $ACCESSTOKEN

Reviewer 2: @nyagamunene 
```bash
magistrala-cli messages read $CHANNELID $ACCESSTOKEN             

error: Status: : Get "http://localhost:9011/channels/e54b2069-9468-49a2-bb48-4a317a13ab33/messages?limit=10&offset=0&total=0": dial tcp 127.0.0.1:9011: connect: connection refused
```

```

   This reads messages from the specified channel.

You can also update a Thing's tags, secret, and metadata. Here's an example of how to do that:

```bash
magistrala-cli things update secret $THINGID $THINGSECRET $ACCESSTOKEN
```

> 

Reviewer 2: ```suggestion
**Note:** Make sure to replace `$THINGID`, `$THINGSECRET`, and `$ACCESSTOKEN` with your actual Thing ID, Thing secret, and access token values, respectively.
```

Within your domain, you have the ability to invite or assign other users as members, guests, contributors, editors, or administrators. This allows you to grant varying levels of permissions to users over different entities. To assign a user to a domain, use the following command:

```bash

> magistrala-cli domains assign users member '[<user_id>]' $DOMAINID $ACCESSTOKEN

Reviewer 2: @nyagamunene please check this

```

- To send a message over HTTP, you’ll need the Channel ID and Thing Secret. The endpoint URL will look like this: `https://localhost/http/channels/{channel_id}/messages`. Here’s how to do it:

```bash
curl --location 'http://localhost/http/channels/c4935742-1422-4636-9442-e7eeb7c8c681/messages' \
--header 'Content-Type: application/senml+json' \
--header 'Authorization: Thing 91f6e004-9f9e-4f95-9b72-a04befbdf584' \
--data '[{"bn": "DS-5000-AL1-001", "n": "Distance_AssemblyLine1", "u": "mm","v": 152.3}, { "n": "BatteryLevel", "u": "%", "v": 95}, { "n": "SignalStrength", "u": "dBm", "v": -65}]'
```

- To read the message, use a similar command:

```bash
curl --location 'http://localhost/http/channels/aecf0902-816d-4e38-a5b3-a1ad9a7cf9e8/messages' \
--header 'Authorization: Thing a83b9afb-9022-4f9e-ba3d-4354a08c273a'
```

> For more information about using HTTP api, refer to the [documentation][swagger-docs] on swagger.

Reviewer 2: ```suggestion

The `--curl` flag in `magistrala-cli`, provides the equivalent `curl` request that the CLI would execute. This can be helpful for debugging or using the command outside of the CLI. For example, when you create a new user with the following command:

```bash
magistrala-cli users create john johndoe@example.com 12345678 --curl
```
You will receive the equivalent `curl` command:

```bash
2024/09/06 15:00:30 curl -X 'POST' -d '{"id":"","name":"john","credentials":{"identity":"johndoe@example.com","secret":"12345678"},"created_at":"0001-01-01T00:00:00Z","updated_at":"0001-01-01T00:00:00Z","status":"enabled"}' -H 'Content-Type: application/json' 'http://localhost:9002/users'
```
The output shows the HTTP request that the CLI would execute, allowing you to copy and paste it for manual execution with `curl`. For more information about using HTTP api, refer to the [documentation][swagger-docs] on swagger.
```

9. **Customizable and Extensible**: Magistrala's architecture allows for easy customization and extension to meet specific business needs.

**Please note:** Magistrala offers a variety of services, each detailed comprehensively in the [architecture documentation][architecture].

## Getting Started with Magistrala

Setting up Magistrala is straightforward. With Docker (version 26.0.0 and above) and Go (version 1.21 and above) installed on your system, you can clone the official GitHub repository and have the platform up and running in no time. Here's a quick guide to get you started:

- Clone the Magistrala repository:

   ```bash
   git clone https://github.com/absmach/magistrala.git
   cd magistrala
   ```

- Build and install the binaries:

   ```bash
   make && make install
   ```

**Note:** This process will compile the binaries into the `<project_root>/build` directory. If the `$GOBIN` environment variable is set, the binaries will also be copied to the `go/bin` directory.

To initiate the Magistrala Docker services, execute the following command from the project's root directory:

- Start the Magistrala Docker services:

   ```bash
   make run
   ```

> The platform offers multiple interaction options, including a CLI, SDK, HTTP API, and UI, catering to different user preferences and use cases.

Reviewer 2: ```suggestion
Magistrala offers multiple interaction options, including a CLI, SDK, HTTP API, and UI, catering to different user preferences and use cases.
```

- Generate an access token:

```bash
magistrala-cli users token johndoe@example.com 12345678
```

   This generates an access token for the user, which is required for authentication in subsequent operations.

   The access token can be set to an environmental variable to allow for easy use:

```bash
ACCESSTOKEN=eyJhbGciOiJIUzUxMiIsInR5cCI6IkpXVCJ9.eyJleHAiOjE3MTgxOTI2MTIsImlhdCI6MTcxODE4OTAxMiwiaXNzIjoibWFnaXN0cmFsYS5hdXRoIiwidHlwZSI6MCwidXNlciI6ImMxYjQ3OWQwLWEzNDYtNGZiYy1hN2FkLWVkNGFiNjcxYTIyZCJ9._Qlri9FN_E2vJLZIVQzyeeg9I_ggVRi_CKS52xlUD8YEkSKWkTlmr02WlKdhvR-aZzaDyudICLtCPtadMCGCzg
```

**Note:** The token last for 15 minutes before expiring and will need the user to get a new one.

- Create a domain (workspace):

```bash
magistrala-cli domain create workspace alias $ACCESSTOKEN
```

   This creates a new domain with the name "workspace" with an alias. The $ACCESSTOKEN is the token generated in the previous step.

Save the domain Id to use it for logging in:

```bash
DOMAINID=9a00bc60-2e07-4503-8144-a24e748175d0
```

To login into your workspace to generate the token with the domain ID attached:

```bash
magistrala-cli users token johndoe@example.com 12345678 $DOMAINID
```

> **Note:** The access token will need to set to the environmental variable created earlier. While logged into your domain there are muliple things that can be done. Let go through some of them.

Reviewer 2: ```suggestion
**Note:** The access token will need to be set to the `ACCESSTOKEN` environmental variable created earlier. While logged into your domain, there are multiple operations that can be done. Let go through some of them.
```
